File: 015.py
#!/usr/bin/env python
"""
Problem 15
19 April 2002
 
Starting in the top left corner of a 2*2 grid, there are 6 routes (without
backtracking) to the bottom right corner.

TODO: Figure out how to insert these images

How many routes are there through a 20*20 grid?

Answer: ?????
"""

map = {}

def routes(x, y):
    """Calculate unique routes across the diagonal of an an x by y grid"""
    if 1 == x and 1 == y:
        return 2
    elif 0 == x or 0 == y:
        return 1
    elif (x,y) in map:
        return map[(x,y)]
    elif (y,x) in map:
        return map[(y,x)]
    else:
        map[(x,y)] = routes(x-1, y) + routes(x, y-1)
        return map[(x,y)]

def solve():
    #print routes(2,2)
    return routes(20,20)

if __name__ == '__main__':
    print "Answer: %s" % solve()

File: Ruby/015.rb
#!/usr/bin/env ruby

# Lattice paths
# Problem 15
#
# Starting in the top left corner of a 2×2 grid, and only being able to move to
# the right and down, there are exactly 6 routes to the bottom right corner.
#
#       <lattice.gif>
#
# How many such routes are there through a 20×20 grid?


Memo = Hash.new

def routes x, y
    if x == 0 || y == 0
        return 1
    elsif x == 1 && y == 1
        return 2
    elsif Memo[[x,y]]
        return Memo[[x,y]]
    else
        Memo[[y,x]] = Memo[[x,y]] = routes(x-1, y) + routes(x, y-1)
        return Memo[[x,y]]
    end
end

def p15
    return routes(20,20)
end


if __FILE__ == $0
    puts p15
end

